lib/charms/vault_k8s/v0/vault_tls.py

```diff
@@ -0,0 +1,451 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""This file includes methods to manage TLS certificates within the Vault charms."""
+
+import logging
+import socket
+from enum import Enum, auto
+from signal import SIGHUP
+from typing import List, Optional, TextIO, Tuple
+
+from charms.certificate_transfer_interface.v0.certificate_transfer import (
+    CertificateTransferProvides,
+)
+from charms.tls_certificates_interface.v2.tls_certificates import (
+    TLSCertificatesRequiresV2,
+    generate_ca,
+    generate_certificate,
+    generate_csr,
+    generate_private_key,
+)
+from ops import EventBase, Object, RelationBrokenEvent, SecretNotFoundError
+from ops.pebble import APIError, PathError
+
+from exceptions import PeerSecretError, VaultCertsError
+
+# The unique Charmhub library identifier, never change it
+LIBID = "61b41a053d9847ce8a14eb02197d12cb"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+logger = logging.getLogger(__name__)
+
+SEND_CA_CERT_RELATION_NAME = "send-ca-cert"
+TLS_CERTIFICATE_ACCESS_RELATION_NAME = "tls-certificates-access"
+TLS_FILE_FOLDER_PATH = "/vault/certs"
+CA_CERTIFICATE_JUJU_SECRET_KEY = "vault-ca-certificates-secret-id"
+CA_CERTIFICATE_JUJU_SECRET_LABEL = "vault-ca-certificate"
+
+VAULT_CA_SUBJECT = "Vault self signed CA"
+
+
+class Substrate(Enum):
+    """Determines which type of charm this library is running in.
+
+    Some library functions need to interact with the substrate by reading and writing files. This
+    enum allows the library to make the correct decision on which methods to use depending on the
+    substrate.
+    """
+
+    KUBERNETES = "kubernetes"
+    MACHINE = "machine"
+
+
+class File(Enum):
+    """This enum determines which files are expected of the library to read."""
+
+    CERT = auto()
+    KEY = auto()
+    CA = auto()
+    CSR = auto()
+
+
+class VaultTLSManager(Object):
+    """This class configures the certificates within Vault."""
+
+    def __init__(self, charm, peer_relation: str, substrate: Substrate):
+        """Manager of MongoDB TLS relation and configuration."""
+        super().__init__(charm, "tls")
+        self.charm = charm
+        self.substrate = substrate
+        self.peer_relation = peer_relation
+        self.tls_access = TLSCertificatesRequiresV2(charm, TLS_CERTIFICATE_ACCESS_RELATION_NAME)
+        self.certificate_transfer = CertificateTransferProvides(charm, SEND_CA_CERT_RELATION_NAME)
```

Comment:
I'm not sure we want this file to handle those 2 integrations as they are quite different. Again, I'm ok with keeping them here at the moment, this can be done in the future if it ends up being a pain.

Reply:
It was pretty common for me to send ca certificates when I replace them within vault, so I moved it here. To me it makes sense because sending the ca cert is relevant to multiple relations like kv and ingress, and is a responsibility of the TLS configuration so I moved it to a place where I expected to find it.

I'm open to any alternatives.